### ISSUE TITLE
Unit tests leave behind open files

### ISSUE THREAD

**Issue by jdavies-st** [spacetelescope/jwst#2887]
The following unit tests leave behind open files:
```
[gw0] [ 10%] ERROR jwst/assign_wcs/tests/test_schemas.py::test_distortion_schema
[gw1] [ 12%] ERROR jwst/datamodels/tests/test_fits.py::test_delete 
[gw1] [ 12%] ERROR jwst/datamodels/tests/test_fits.py::test_extra_fits 
[gw1] [ 12%] ERROR jwst/datamodels/tests/test_fits.py::test_casting 
[gw1] [ 12%] ERROR jwst/datamodels/tests/test_fits.py::test_metadata_doesnt_override 
[gw1] [ 12%] ERROR jwst/datamodels/tests/test_fits.py::test_metadata_from_fits 
[gw1] [ 13%] ERROR jwst/datamodels/tests/test_history.py::test_history_from_model_to_fits 
[gw1] [ 13%] ERROR jwst/datamodels/tests/test_history.py::test_history_from_fits 
[gw1] [ 14%] ERROR jwst/datamodels/tests/test_models.py::test_subarray 
[gw1] [ 15%] ERROR jwst/datamodels/tests/test_models.py::test_from_fits_write 
[gw1] [ 16%] ERROR jwst/datamodels/tests/test_models.py::test_multislit_from_fits_image 
[gw1] [ 16%] ERROR jwst/datamodels/tests/test_models.py::test_multislit_copy 
[gw1] [ 18%] ERROR jwst/datamodels/tests/test_models.py::test_image_with_extra_keyword_to_multislit 
[gw1] [ 19%] ERROR jwst/datamodels/tests/test_open.py::test_open_image 
[gw1] [ 19%] ERROR jwst/datamodels/tests/test_open.py::test_open_reference_files 
[gw1] [ 20%] ERROR jwst/datamodels/tests/test_schema.py::test_mask_model 
[gw1] [ 21%] ERROR jwst/datamodels/tests/test_schema.py::test_validate_transform_from_file 
[gw3] [ 33%] ERROR jwst/datamodels/tests/test_fits.py::test_from_fits 
[gw3] [ 33%] ERROR jwst/datamodels/tests/test_fits.py::test_from_scratch 
[gw3] [ 42%] ERROR jwst/lib/tests/test_set_telescope_pointing.py::test_add_wcs_default 
[gw3] [ 42%] ERROR jwst/lib/tests/test_set_telescope_pointing.py::test_add_wcs_fsmcorr_v1 
[gw3] [ 42%] ERROR jwst/lib/tests/test_set_telescope_pointing.py::test_add_wcs_with_db 
[gw3] [ 42%] ERROR jwst/lib/tests/test_set_telescope_pointing.py::test_add_wcs_with_db_fsmcorr_v1 
[gw3] [ 65%] ERROR jwst/lib/tests/test_utc_to_tdb.py::test_utc_to_tdb 
[gw2] [ 68%] ERROR jwst/datamodels/tests/test_wcs.py::test_wcs 
[gw3] [ 71%] ERROR jwst/resample/tests/test_interface.py::test_multi_integration_input[ResampleSpecStep] 
[gw3] [ 73%] ERROR jwst/stpipe/tests/test_input.py::test_default_input_with_container 
[gw3] [ 79%] ERROR jwst/stpipe/tests/test_logger.py::test_configuration 
[gw3] [ 80%] ERROR jwst/stpipe/tests/test_pipeline.py::test_pipeline 
[gw0] [ 80%] ERROR jwst/stpipe/tests/test_saving.py::test_save_substep_withdir 
[gw2] [ 80%] ERROR jwst/dq_init/tests/test_dq_init.py::test_fullstep[GuiderRawModel-Image] 
[gw3] [ 81%] ERROR jwst/stpipe/tests/test_pipeline.py::test_pipeline_commandline 
[gw3] [ 82%] ERROR jwst/stpipe/tests/test_saving.py::test_save_pipeline_default 
[gw3] [ 82%] ERROR jwst/stpipe/tests/test_saving.py::test_save_pipeline_withdir 
[gw2] [ 97%] ERROR jwst/exp_to_source/tests/test_exp_to_source.py::test_model_structure 
```
It's not clear whether the test itself or the library code being tested is leaving the open files.  But we should fix them.

FWIW, the way to find these is to use the `pytest-openfiles` plugin available via `pip` or `conda`:
```
pytest --open-files
```

**Comment by jdavies-st** [comment]
Ok, so at least one of these are because of the way the test is written.  But many are due to what seems to be a deeper problem in `jwst.datamodels`.  Here's a test case:
```
In [1]: from jwst import datamodels 
   ...: import psutil 
   ...: proc = psutil.Process()                                                                                                                                 

In [2]: print(proc.open_files())                                                                                                                                
[popenfile(path='/Users/jdavies/.ipython/profile_default/history.sqlite', fd=5), popenfile(path='/Users/jdavies/.ipython/profile_default/history.sqlite', fd=6)]

In [3]: print(proc.open_files()[2:])                                                                                                                            
[]

In [4]: with datamodels.RampModel('jwst/datamodels/tests/data/test.fits') as dm: 
   ...:     pass 
   ...:                                                                                                                                                         

In [5]: print(proc.open_files()[2:])                                                                                                                            
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=14)]

In [6]: with datamodels.RampModel('jwst/datamodels/tests/data/test.fits') as dm: 
   ...:     pass 
   ...:                                                                                                                                                         

In [7]: print(proc.open_files()[2:])                                                                                                                            
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=14), popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=15)]

In [8]: with datamodels.RampModel('jwst/datamodels/tests/data/test.fits') as dm: 
   ...:     pass 
   ...:                                                                                                                                                         

In [9]: print(proc.open_files()[2:])                                                                                                                            
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=13), popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=14), popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=15)]

In [10]: dm.close()                                                                                                                                             

In [11]: print(proc.open_files()[2:])                                                                                                                           
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=13), popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=14), popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=15)]

In [18]: import gc                                                                                                                                              

In [19]: gc.collect()                                                                                                                                           
Out[19]: 804

In [20]: print(proc.open_files()[2:])                                                                                                                           
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=13)]

In [21]: dm.close()                                                                                                                                             

In [22]: print(proc.open_files()[2:])                                                                                                                           
[popenfile(path='/Users/jdavies/dev/jwst/jwst/datamodels/tests/data/test.fits', fd=13)]

In [27]: dm                                                                                                                                                     
Out[27]: <RampModel(5, 35, 40, 32) from test.fits>
```
So not only does `datamodels` not respect the `with` context manager, but it also doesn't respect the `close()` method.

@stscieisenhamer any ideas?

**Comment by jdavies-st** [comment]
I feel like we had solved this in #2360, but here it is again.

**Comment by philhodge** [comment]
I tried an asdf file:

dm = datamodels.open("jwst_nirspec_fpa_0005.asdf")
dm
<DataModel from jwst_nirspec_fpa_0005.asdf>
dm.close()
dm
<DataModel from jwst_nirspec_fpa_0005.asdf>
but:
print(proc.open_files())
[]

So maybe the problem is in `astropy.io.fits`?

**Comment by stscieisenhamer** [comment]
Hmmm, I thought there was a caveat comment on #2360 or initial issue #2282, but don't see it so here it is.

#2360 did solve the file handle/reference leak. However, it should also be noted that with memmap, even if a garbage collection is done, file handles can remain alive.  In so much as leaking, the file handles do get reaped, so that when a process runs long and opens/closes many files, the reaping is enough to not cause resource exhaustion. Hence, the issue was resolved.

Whether the keep-alive of the file handles is an issue deeper in astropy.io.fits in how it handles memmap, or if this is a "feature" of the os itself is unknown. Regardless. the presumption is that whatever is checking for open files must be doing so in the same process that `pytest` is running in. Another presumption is that the file handle check is being done at the os-level. If these presumptions are true, then it is of no surprise that it is finding open files. However, at this time, these are best counted as "false positives": Though the handles exist, the logic that opened them have also closed them, at the language (Python) level. More specifically, `datamodels` and `asdf` were cleared of any wrong-doing as related to #2282/#2360.

**Comment by stscieisenhamer** [comment]
Also, a distinction of "file handle" need be made: There are two in this context, the Python `file` object, and the OS-level `file handle`. #2360 resolved the Python `file` leak. The above comment is referring to the keep-alive of the OS-level file handles themselves, and the presumption that whatever is doing the check is checking OS-level file handles. If the check is actually looking at Python `file`s, then there is a bonafide Python file leak still remaining.

**Comment by jdavies-st** [comment]
The check is looking at Python files I believe, but perhaps @drdavella can clarify?  I believe the `pytest-openfiles` plugin actually uses the same `psutil.Process.open_files()` method as I used above in the simple test case.

**Comment by jdavies-st** [comment]
I guess the thing that bugs me is that `datamodels` do not respect context managers.  I.e. I can access the datamodel after I have left the `with` context.

Now I know there's something related here with `memmap` and FITS files, as the `astropy.io.fits` library doesn't respect them either.  I don't know how this affects `datamodels` though.

**Comment by drdavella** [comment]
@jdavies-st is right: `pytest-openfiles` is just checking the list of files using `psutil` before and after running each test.

Also agree with @jdavies-st here: there's no point of using a context manager if it doesn't actually manage the underlying resources. There's not much we can do about FITS, but it seems like datamodels can handle things properly.

**Comment by stscieisenhamer** [comment]
`psutil.Process.open_files` is referring to OS-level file handles, or more *NIX proper, file descriptors.

WRT context management: Agreed that that needs fixing or not be allowed.

**Comment by jdavies-st** [comment]
Thank for the extra info @stscieisenhamer and @drdavella.

The background here is that I'm trying parallelize our reg test runs.  It works and reduces run time to 1.5 hours, as jwcalibdev is a beast.  But the AMI tests are causing `xdist` to fail, and they uniquely, get a files not closed warning when run under `pytest`.  So that's the current motivation, besides context managers needing to work as well.

**Comment by jdavies-st** [comment]
This was fixed by #4445 

Now:

```
In [1]: from jwst import datamodels

In [2]: import psutil

In [3]: proc = psutil.Process()

In [6]: print(proc.open_files()[2:])
[]

In [7]: with datamodels.RampModel('jwst/datamodels/tests/data/test.fits') as dm:
   ...:     pass 
   ...:                                                                                                                                         

In [8]: print(proc.open_files()[2:])
[]
```

No files left open.
